### PR TITLE
[WIP] finish up the get_output conversion (2nd attempt)

### PR DESCRIPTION
**Pull request by benanne**
This time without the rebase! I'll see if I can update the tests / examples accordingly as well.

### CODE REVIEW

**Comment by benanne** [comment]
I just made `Layer.get_output()` and `Layer.get_output_shape()` raise a `RuntimeError` to figure out where these methods are still being used. I'm getting 26 test failures as a result (with `--runslow`) so I'll try to fix those and then remove the raise statements again.

**Comment by benanne** [comment]
I fixed a bunch of them, there are still some issues with `test_objectives.py`, but since that module is slated for being overhauled soon anyway I'm not going to bother to fix them.

**Comment by benanne** [comment]
By the way, it might be a better idea to cherry-pick / merge this and make the changes you want rather than to wait for me to do it, I should go back to writing for a bit now :)

**Comment by f0k** [comment]
Hey, thanks! Just came back from Athens and was going to do this, nice to see you already did most of the work :) I'll have a look and merge as I see fit.

**Comment by benanne** [comment]
Sounds good, hope you enjoyed Athens!

File: lasagne/layers/helper.py [hunk 1]
```diff
@@ -233,17 +233,66 @@ def get_output_shape(layer_or_layers):
...
+                if isinstance(layer, MultipleInputsLayer):
+                    input_shapes = [all_shapes[input_layer]
+                                    for input_layer in layer.input_layers]
+                else:
+                    input_shapes = all_shapes[layer.input_layer]
+            except KeyError:
+                raise ValueError("get_output() was called without giving an "
+                                 "input shape for the free-floating layer %r. "
+                                 "Please call it with a dictionary mapping "
+                                 "this layer to an input expression."
```
Comment (f0k) [inline lasagne/layers/helper.py]: note to myself: change this to "input shape"

File: lasagne/tests/layers/test_base.py [hunk 2]
```diff
@@ -10,8 +10,15 @@ def layer(self):
         from lasagne.layers.base import Layer
         return Layer(Mock())
```
Comment (f0k) [inline lasagne/tests/layers/test_base.py]: note to myself: This fixture is unused now. Change or remove it.

**Pull request merged**